File: java/lc208.java
class Trie {
    boolean wordEnd;
    Trie[] child;
    /** Initialize your data structure here. */
    public Trie() {
        wordEnd = false;
        child = new Trie[26];
        for(int i=0;i<26;i++)
        {
            child[i] = null;
        }    
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
         Trie t = this;
        for(int i=0;i<word.length();i++)
        {
            if(t.child[word.charAt(i)-'a']==null)
            {
                t.child[word.charAt(i)-'a']= new Trie();
            }
            t = t.child[word.charAt(i)-'a'];
        }
        t.wordEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
                Trie t = this;
        for(int i=0;i<word.length();i++)
        {
            if(t.child[word.charAt(i)-'a']==null)
            {
                return false;
            }
            t = t.child[word.charAt(i)-'a'];
        }
        return t.wordEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
                Trie t = this;
        for(int i=0;i<prefix.length();i++)
        {
            if(t.child[prefix.charAt(i)-'a']==null)
            {
                return false;
            }
            t = t.child[prefix.charAt(i)-'a'];
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: python/lc208.py
class Trie:

    def __init__(self):
        self.eow = False
        self.items = [None]*26
        

    def insert(self, word: str) -> None:
        idx = ord(word[0])-ord('a')
        if not self.items[idx]:
            self.items[idx] = Trie()
        if len(word) > 1:
            self.items[idx].insert(word[1:])
        else:
            self.items[idx].eow = True

    def search(self, word: str) -> bool:
        idx = ord(word[0])-ord('a')
        if not self.items[idx]:
            return False
        else:
            if len(word) == 1:
                return self.items[idx].eow
            else:
                return self.items[idx].search(word[1:]) 

    def startsWith(self, prefix: str) -> bool:
        idx = ord(prefix[0])-ord('a')
        if not self.items[idx]:
            return False
        else:
            if len(prefix) == 1:
                return True
            else:
                return self.items[idx].startsWith(prefix[1:])



# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
